paddle/fluid/framework/paddle2cinn/cinn_subgraph_detector.cc

```diff
@@ -0,0 +1,352 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/framework/paddle2cinn/cinn_subgraph_detector.h"
+#include "glog/logging.h"
+#include "paddle/fluid/framework/ir/subgraph_detector.h"
+
+namespace paddle {
+namespace framework {
+namespace paddle2cinn {
+
+std::unordered_set<Node*> GetProducerOps(Node* node) {
+  CHECK(node->IsOp());
+  std::unordered_set<Node*> producers;
+
+  for (auto input_var : node->inputs) {
+    CHECK(input_var->IsVar());
+    for (auto input_op : input_var->inputs) {
+      CHECK(input_op->IsOp());
+      producers.insert(input_op);
+    }
+  }
+
+  return producers;
+}
+
+std::unordered_set<Node*> GetConsumerOps(Node* node) {
+  CHECK(node->IsOp());
+  std::unordered_set<Node*> consumers;
+
+  for (auto output_var : node->outputs) {
+    CHECK(output_var->IsVar());
+    for (auto output_op : output_var->outputs) {
+      CHECK(output_op->IsOp());
+      consumers.insert(output_op);
+    }
+  }
+
+  return consumers;
+}
+
+struct Hasher {
+  size_t operator()(const CinnSubGraphPtr& subgraph) const noexcept {
+    return std::hash<uint64_t>()(reinterpret_cast<uint64_t>(subgraph.get()));
```

Comment:
写这么复杂其实没必要，直接
```suggestion
    return std::hash<CinnSubGraphPtr>()(subgraph);
```
就行了，`std::hash`是支持了`shared_ptr `的：https://zh.cppreference.com/w/cpp/memory/shared_ptr/hash

不过问题不大，不修改也没问题。